lib/ansible/plugins/loader.py

```diff
@@ -220,16 +228,16 @@ def _load_config_defs(self, name, module, path):
     def add_directory(self, directory, with_subdir=False):
         ''' Adds an additional directory to the search path '''
 
-        directory = os.path.realpath(directory)
+        directory = os.path.realpath(to_bytes(directory))
 
         if directory is not None:
             if with_subdir:
-                directory = os.path.join(directory, self.subdir)
+                directory = os.path.join(directory, to_bytes(self.subdir))
             if directory not in self._extra_dirs:
                 # append the directory and invalidate the path cache
                 self._extra_dirs.append(directory)
                 self._paths = None
-                display.debug('Added %s to loader search path' % (directory))
+                display.debug(u'Added %s to loader search path' % (to_text(directory)))
 
     def _find_plugin(self, name, mod_type='', ignore_deprecated=False, check_aliases=False):
         ''' Find a plugin named name '''
```

Comment:
as part of cleaning up the unicode consistency of the cache keys, we should ensure that `name` and `mod_type` are both `to_text()`ed right at the top of this method, otherwise we could potentially have bogus cache misses under Py3 (since we look up based on whatever came in, but we're now always adding entries as text per the changes below)